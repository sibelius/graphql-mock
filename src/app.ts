import Koa from 'koa';
import logger from 'koa-logger';
import Router from 'koa-router';
import cors from 'kcors';
import graphqlHttp from 'koa-graphql';
import bodyParser from 'koa-bodyparser';
import { makeExecutableSchema, addMocksToSchema } from 'graphql-tools';

const app = new Koa();

const router = new Router();

app.use(bodyParser());
app.on('error', err => {
  // eslint-disable-next-line
  console.log('app error: ', err);
});

app.use(logger());
app.use(cors());

router.get('/', async ctx => {
  ctx.body = 'Welcome to GraphQL Mock server';
});

app.use(async (ctx, next) => {
  const { schema, mocks = {} } = ctx.request.body;

  if (!schema) {
    ctx.status = 400;
    ctx.body = 'missing schema on body parameter'
    return;
  }

  ctx.typeDefs = schema;
  ctx.mocks = mocks;

  await next();
})

router.all('/graphql', graphqlHttp(async (request: Request, ctx: Response, koaContext: KoaContext) => {
  const { typeDefs, mocks } = koaContext;

  const schema = makeExecutableSchema({ typeDefs });

  addMocksToSchema({ schema, mocks });

  return {
    graphiql: true,
    schema,
    formatError: (error) => {
      return {
        message: error.message,
        locations: error.locations,
        stack: error.stack,
      };
    }
  }
}));

app.use(router.routes()).use(router.allowedMethods());

export default app;