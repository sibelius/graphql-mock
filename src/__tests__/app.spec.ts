import request from 'supertest';
import app from '../app';

it('should return query from schema and mocks', async () => {
  const typeDefs = `
type Query {
  user: User
}

type User {
  name: String
}
`;

  const mocks = {
    Query: () => ({
      user: {
        name: 'awesome',
      },
    }),
    User: () => ({
      name: 'Sibelius',
    }),
  }

  // language=GraphQL
  const query = `
  query Q {
    user {
      name
    }   
  }
`;

  const variables = {
  };

  const payload = {
    query,
    variables,
    schema: typeDefs,
    mocks,
  };

  const response = await request(app.callback())
    .post('/graphql')
    .set({
      Accept: 'application/json',
      'Content-Type': 'application/json',
    })
    .send(JSON.stringify(payload));

  expect(response.status).toBe(200);
  expect(response.body).toMatchSnapshot();
});