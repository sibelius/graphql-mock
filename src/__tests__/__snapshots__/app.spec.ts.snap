// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return query from schema and mocks 1`] = `
Object {
  "data": Object {
    "user": null,
  },
}
`;
